benchmarks/missing_type_caching/benchmarker.rb

```diff
@@ -0,0 +1,63 @@
+require 'erb'
+require 'ostruct'
+require 'fileutils'
+require 'json'
+
+class Benchmarker
+  include FileUtils
+
+  def initialize(target, size)
+    @target = target
+    @size = size
+  end
+
+  def setup
+    require 'puppet'
+    config = File.join(@target, 'puppet.conf')
+    Puppet.initialize_settings(['--config', config])
+    Puppet[:always_retry_plugins] = false
+  end
+
+  def run(args=nil)
+    env = Puppet.lookup(:environments).get('benchmarking')
+    node = Puppet::Node.new("testing", :environment => env)
+    Puppet::Resource::Catalog.indirection.find("testing", :use_node => node)
+  end
+
+  def generate
+    environment = File.join(@target, 'environments', 'benchmarking')
+    templates = File.join('benchmarks', 'missing_type_caching')
+    test_module_dir = File.join(environment, 'modules', 'testmodule',
+                                'manifests')
+
+    mkdir_p(File.join(environment, 'modules'))
+    @size.times.each do |i|
```

Comment:
These extra directories aren't strictly necessary, but they make the performance gains more apparent. Without these lines, the following results are achieved.

Default setting value (true, no miss caching):
```
 SIZE=500 bundle exec rake benchmark:missing_type_caching:run
mkdir -p /var/folders/zm/nk5zs3kj1zjdwf3mjcdjs19m0000gq/T/missing_type_caching20160404-7358-1ghwvae
                 user     system      total        real
Run 1        0.600000   0.100000   0.720000 (  0.714323)
Run 2        0.570000   0.090000   0.660000 (  0.665010)
Run 3        0.620000   0.090000   0.710000 (  0.706283)
Run 4        0.520000   0.090000   0.610000 (  0.609977)
Run 5        0.570000   0.090000   0.660000 (  0.678567)
Run 6        0.550000   0.090000   0.640000 (  0.633102)
Run 7        0.560000   0.090000   0.650000 (  0.649624)
Run 8        0.590000   0.080000   0.670000 (  0.667839)
Run 9        0.560000   0.090000   0.650000 (  0.651638)
Run 10       0.640000   0.080000   0.720000 (  0.722062)
> total:     5.780000   0.890000   6.690000 (  6.698426)
> avg:       0.578000   0.089000   0.669000 (  0.669843)
```

always_retry_plugins set to false (caching enabled):
```
SIZE=500 bundle exec rake benchmark:missing_type_caching:run
mkdir -p /var/folders/zm/nk5zs3kj1zjdwf3mjcdjs19m0000gq/T/missing_type_caching20160404-7451-17f2ypv
                 user     system      total        real
Run 1        0.340000   0.020000   0.360000 (  0.372802)
Run 2        0.340000   0.000000   0.340000 (  0.347129)
Run 3        0.290000   0.010000   0.300000 (  0.290051)
Run 4        0.250000   0.000000   0.250000 (  0.263602)
Run 5        0.280000   0.010000   0.290000 (  0.276459)
Run 6        0.300000   0.000000   0.300000 (  0.304726)
Run 7        0.330000   0.000000   0.330000 (  0.334759)
Run 8        0.300000   0.000000   0.300000 (  0.303248)
Run 9        0.260000   0.010000   0.270000 (  0.266815)
Run 10       0.250000   0.010000   0.260000 (  0.253278)
> total:     2.940000   0.060000   3.000000 (  3.012867)
> avg:       0.294000   0.006000   0.300000 (  0.301287)
```